File: Leetcode_Code/230. Kth Smallest Element in a BST/230.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode *root, int k) {
        int count = k, mini = INT_MIN;
        kthSmallest_dfs(root, count, mini);
        return mini;
    }

private:
    void kthSmallest_dfs(TreeNode *root, int &count, int &mini) {
        if (!root) return;
        if (count > 0)
            kthSmallest_dfs(root->left, count, mini);
        if (mini <= root->val && count > 0) {
            count--;
            mini = root->val;
        }
        if (count > 0)
            kthSmallest_dfs(root->right, count, mini);

    }
};

File: Leetcode_Code/230. Kth Smallest Element in a BST/230.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        count = None
        mini = None

    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
        self.count = k
        self.kthSmallest_dfs(root)
        return self.mini

    def kthSmallest_dfs(self, root):
        if root is None:
            return
        if self.count > 0:
            self.kthSmallest_dfs(root.left)
        if self.count > 0:
            self.count -= 1
            self.mini = root.val
        if self.count > 0:
            self.kthSmallest_dfs(root.right)
